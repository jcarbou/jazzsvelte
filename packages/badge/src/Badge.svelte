<script lang="ts">
    import type { CssStyle, HTMLSpanAttributes } from '@jazzsvelte/api'
    import type {
        BadgePassThroughMethodOptions,
        BadgePassThroughOptions,
        BadgeSeverity,
        BadgeSize,
        BadgeValue
    } from './badge.types'

    import { JAZZ_SVELTE, resolvePT } from '@jazzsvelte/api'
    import { isEmpty, isNotEmpty } from '@jazzsvelte/object'

    export let value: BadgeValue = null
    export let severity: BadgeSeverity = null
    export let size: BadgeSize = null
    export let pt: BadgePassThroughOptions | null = null
    export let ptOptions: BadgePassThroughMethodOptions | null = null
    export let unstyled: boolean = false
    export let style: CssStyle = null
    let className: string | null = null
    export { className as class }

    // "root element"
    $: rootAttributes = resolvePT(
        {
            class: [
                'p-badge p-component',
                className,
                {
                    'p-badge-no-gutter': isNotEmpty(value) && String(value).length === 1,
                    'p-badge-dot': isEmpty(value),
                    'p-badge-m': size === 'medium',
                    'p-badge-lg': size === 'large',
                    'p-badge-xl': size === 'xlarge',
                    [`p-badge-${severity}`]: severity !== null
                }
            ],
            style,
            'data-pc-name': 'badge',
            'data-pc-section': 'root'
        },
        pt?.root,
        JAZZ_SVELTE.pt?.badge?.root,
        ptOptions,
        unstyled
    ) satisfies HTMLSpanAttributes
</script>

<span {...rootAttributes} {...$$restProps}>{value}</span>

<svelte:head>
    <style type="text/css">
        .p-overlay-badge {
            position: relative;
        }

        .p-overlay-badge .p-badge {
            position: absolute;
            top: 0;
            right: 0;
            transform: translate(50%, -50%);
            transform-origin: 100% 0;
            margin: 0;
        }
    </style>
</svelte:head>

<style>
    @layer primereact {
        .p-badge {
            display: inline-block;
            border-radius: 10px;
            text-align: center;
            padding: 0 0.5rem;
        }

        .p-badge.p-badge-m {
            font-size: 0.725rem;
            min-width: 1.55rem;
            height: 1.55rem;
            line-height: 1.55rem;
        }

        .p-badge-dot {
            width: 0.5rem;
            min-width: 0.5rem;
            height: 0.5rem;
            border-radius: 50%;
            padding: 0;
        }

        .p-badge-no-gutter {
            padding: 0;
            border-radius: 50%;
        }
    }
</style>
