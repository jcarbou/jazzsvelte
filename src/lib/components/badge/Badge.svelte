<script lang="ts">
    import type { CssStyle } from '../utils/cssStyles'
    import type {
        BadgePassThroughMethodOptions,
        BadgePassThroughOptions,
        BadgeSeverity,
        BadgeSize,
        BadgeValue
    } from './badge.types'
    import JAZZ_SVELTE from '../api/JazzSvelte'
    import { resolvePT } from '../utils/ptUtils'
    import { isEmpty, isNotEmpty } from '../utils/Utils'

    export let value: BadgeValue = null
    export let severity: BadgeSeverity = null
    export let size: BadgeSize = null
    export let pt: BadgePassThroughOptions | null = null
    export let ptOptions: BadgePassThroughMethodOptions | null = null
    export let unstyled: boolean = false
    export let style: CssStyle = null

    // "root element"
    $: rootAttributes = resolvePT(
        [
            'p-badge p-component',
            {
                'p-badge-no-gutter': isNotEmpty(value) && String(value).length === 1,
                'p-badge-dot': isEmpty(value),
                'p-badge-m': size === 'medium',
                'p-badge-lg': size === 'large',
                'p-badge-xl': size === 'xlarge',
                [`p-badge-${severity}`]: severity !== null
            }
        ],
        style,
        pt?.root,
        JAZZ_SVELTE.pt?.badge?.root,
        ptOptions,
        unstyled
    )
</script>

<span data-pc-name="badge" data-pc-section="root" {...rootAttributes} {...$$restProps}>{value}</span>

<style>
    @layer primereact {
        .p-badge {
            display: inline-block;
            border-radius: 10px;
            text-align: center;
            padding: 0 0.5rem;
        }

        .p-badge.p-badge-m {
            font-size: 0.725rem;
            min-width: 1.55rem;
            height: 1.55rem;
            line-height: 1.55rem;
        }

        .p-overlay-badge {
            position: relative;
        }

        .p-overlay-badge .p-badge {
            position: absolute;
            top: 0;
            right: 0;
            transform: translate(50%, -50%);
            transform-origin: 100% 0;
            margin: 0;
        }

        .p-badge-dot {
            width: 0.5rem;
            min-width: 0.5rem;
            height: 0.5rem;
            border-radius: 50%;
            padding: 0;
        }

        .p-badge-no-gutter {
            padding: 0;
            border-radius: 50%;
        }
    }
</style>
