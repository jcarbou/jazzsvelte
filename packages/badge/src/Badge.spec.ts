import '@testing-library/jest-dom'
import { describe } from 'vitest'
import Badge from './Badge.svelte'
import { snapshot } from '@jazzsvelte/test'

describe('Badge', () => {
    snapshot('default', Badge, {})
    snapshot('value', Badge, { value: 22 })
    snapshot('size large', Badge, { size: 'large' })
    snapshot('size xlarge', Badge, { size: 'xlarge' })
    snapshot('severity success', Badge, { severity: 'success' })
    snapshot('severity info', Badge, { severity: 'info' })
    snapshot('severity warning', Badge, { severity: 'warning' })
    snapshot('severity danger', Badge, { severity: 'danger' })
})
