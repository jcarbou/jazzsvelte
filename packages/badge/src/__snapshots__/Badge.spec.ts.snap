// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge > default 1`] = `
<span
  class="p-badge p-component p-badge-dot s-8HMn8itYTtQw"
  data-pc-name="badge"
  data-pc-section="root"
  data-tested-element="true"
>
  null
</span>
`;

exports[`Badge > severity danger 1`] = `
<span
  class="p-badge p-component p-badge-dot p-badge-danger s-8HMn8itYTtQw"
  data-pc-name="badge"
  data-pc-section="root"
  data-tested-element="true"
>
  null
</span>
`;

exports[`Badge > severity info 1`] = `
<span
  class="p-badge p-component p-badge-dot p-badge-info s-8HMn8itYTtQw"
  data-pc-name="badge"
  data-pc-section="root"
  data-tested-element="true"
>
  null
</span>
`;

exports[`Badge > severity success 1`] = `
<span
  class="p-badge p-component p-badge-dot p-badge-success s-8HMn8itYTtQw"
  data-pc-name="badge"
  data-pc-section="root"
  data-tested-element="true"
>
  null
</span>
`;

exports[`Badge > severity warning 1`] = `
<span
  class="p-badge p-component p-badge-dot p-badge-warning s-8HMn8itYTtQw"
  data-pc-name="badge"
  data-pc-section="root"
  data-tested-element="true"
>
  null
</span>
`;

exports[`Badge > size large 1`] = `
<span
  class="p-badge p-component p-badge-dot p-badge-lg s-8HMn8itYTtQw"
  data-pc-name="badge"
  data-pc-section="root"
  data-tested-element="true"
>
  null
</span>
`;

exports[`Badge > size xlarge 1`] = `
<span
  class="p-badge p-component p-badge-dot p-badge-xl s-8HMn8itYTtQw"
  data-pc-name="badge"
  data-pc-section="root"
  data-tested-element="true"
>
  null
</span>
`;

exports[`Badge > value 1`] = `
<span
  class="p-badge p-component s-8HMn8itYTtQw"
  data-pc-name="badge"
  data-pc-section="root"
  data-tested-element="true"
>
  22
</span>
`;
